File: 129i-vectors/vectorOps.cpp
/*
 * vectorOps.cpp
 * 
 * Copyright 2013 Michael Wildman <michael@timon>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
using namespace std;

float getLength(vector<float> vect);
void printNorm(vector<float> vect);
void printDotProd(vector<float> vect1, vector<float> vect2);

int main(int argc, char **argv) 
{
	if (argc < 2) return 1;
	int n, d, index;
	
	ifstream inf;
	inf.open(argv[1]);
	if (!inf.good()) return 2;
	
	inf >> n;
	vector< vector<float> > vectors;
	vectors.reserve(n);
	
	for (int i = 0; i < n; i++) {
		inf >> d;
		vectors[i] = vector<float> (d);
		for (int j = 0; j < d; j++) {
			inf >> vectors[i][j];
		}
	}
	inf >> n;
	while (n > 0) {
		char op;
		inf >> op;
		switch (op) {
			case 'l': inf >> index; 
				printf("%.5f\n", getLength(vectors[index]));
				break;
			case 'n': inf >> index;
				printNorm(vectors[index]);
				break;
			case 'd': int index2;
				inf >> index >> index2;
				printDotProd(vectors[index], vectors[index2]);
				break;
			default: return 3;
		}
		n--;
	}
	inf.close();	
	return 0;
}

float getLength(vector<float> vect) {
	float result = 0.0;
	for (float i : vect) result += i * i;
	return sqrt(result);
}

void printNorm(vector<float> vect) {
	float denom = getLength(vect);
	for (float i : vect) printf("%.5f ", i / denom);
	cout << endl;
}

void printDotProd(vector<float> vect1, vector<float> vect2) {
	if (vect1.size() != vect2.size()) cout << "vectors must be same length\n";
	float result = 0.0;
	for (int i = 0; i < vect1.size(); i++) result += vect1[i] * vect2[i];
	printf("%.5f\n", result);
}

File: 129i-vectors/vectorOps.java
/*
 * vectorOps.java
 * 
 * Michael Wildman 2013 <@mwildman.co.nz>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

import java.io.*;
import java.util.Scanner;
import java.text.DecimalFormat;


public class vectorOps {
	public static DecimalFormat df = new DecimalFormat("#.#####");

	public static void main (String args[]) throws IOException {		
		if (args.length > 1) System.exit(1);
		Scanner s = null;
		
		try {
			s = new Scanner(new BufferedReader(new FileReader(args[0])));
			int n = Integer.parseInt(s.next());
			double[][] vectors = new double[n][];
			
			for (int i = 0; i < n; i++) {
				int dimension = Integer.parseInt(s.next());
				vectors[i] = new double[dimension];
				for (int j = 0; j < dimension; j++) 
					vectors[i][j] = Double.parseDouble(s.next());
			}
			n = Integer.parseInt(s.next());
			
			for (int i = 0; i < n; i++) {
				switch (s.next()) {
					case "l": System.out.println(df.format(getLength(vectors[Integer.parseInt(s.next())]))); 
						break;
					case "n": printNormalized(vectors[Integer.parseInt(s.next())]);
						break;
					case "d": printDotProd(vectors[Integer.parseInt(s.next())], vectors[Integer.parseInt(s.next())]);
						break;
					default: System.out.println("input format error");
				}
			} 
		} finally {
			if (s != null) s.close();
		}
	}
	
	public static double getLength(double[] vector) {
		double result = 0.0;
		for (double i : vector)
			result += i * i;
		return Math.sqrt(result);
	}
	
	public static void printNormalized(double[] vector) {
		double length = getLength(vector);
		for (double i : vector)
			System.out.print(df.format(i / length) + " ");
		System.out.println();
	}
	
	public static void printDotProd(double[] vector1, double[] vector2){
		if (vector1.length != vector2.length) System.out.println("vectors must have same dimension");
		else {
			double result = 0.0;
			for (int i = 0; i < vector1.length; i++) {
				result += vector1[i] * vector2[i];
			}
			System.out.println(df.format(result));
		}
	}
}

